core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,275 @@
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	unfinalizedTxs, err := f.txStore.FindConfirmedTxes(ctx, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*Tx
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[common.Hash][]*Tx)
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range unfinalizedTxs {
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			f.lggr.AssumptionViolationw("invalid receipt found for confirmed transaction", "tx", tx, "receipt", receipt)
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
+			receiptBlockHashToTx[receipt.GetBlockHash()] = append(receiptBlockHashToTx[receipt.GetBlockHash()], tx)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.GetBlockNumber().Int64())
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.GetBlockHash().String() {
+			continue
+		}
+		finalizedTxs = append(finalizedTxs, tx)
+	}
+
+	// Check if block hashes exist for receipts on-chain older than the earliest cached head
+	// Transactions are grouped by their receipt block hash to avoid repeat requests on the same hash in case transactions were confirmed in the same block
+	validatedReceiptTxs, err := f.batchCheckReceiptHashesOnchain(ctx, receiptBlockHashToTx, latestFinalizedHead.BlockNumber())
+	if err != nil {
+		// Do not error out to allow transactions that did not need RPC validation to still be marked as finalized
+		// The transactions failed to be validated will be checked again in the next round
+		f.lggr.Errorf("failed to validate receipt block hashes over RPC: %v", err)
+	}
+	finalizedTxs = append(finalizedTxs, validatedReceiptTxs...)
+
+	etxIDs := f.buildTxIdList(finalizedTxs)
+
+	err = f.txStore.UpdateTxesFinalized(ctx, etxIDs, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to update transactions as finalized: %w", err)
+	}
+	return nil
+}
+
+func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, receiptMap map[common.Hash][]*Tx, latestFinalizedBlockNum int64) ([]*Tx, error) {
+	if len(receiptMap) == 0 {
+		return nil, nil
+	}
+	// Group the RPC batch calls in groups of rpcBatchSize
+	var rpcBatchGroups [][]rpc.BatchElem
+	var rpcBatch []rpc.BatchElem
+	for hash := range receiptMap {
+		elem := rpc.BatchElem{
+			Method: "eth_getBlockByHash",
```

Comment:
What additional benefit does this eth_getBlockByHash give us?
I mean, we know that if previously a receipt was seen, then the blockhash for that should be valid.
This call will just re-iterate the same thing.
However we are not sure if that blockhash belongs to the finalized chain. Maybe it was in a fork that didn't make it to the chain.
I think we should validate that, no?

Reply:
We only really fall back to this RPC call if the receipt is older than the earliest head we have cached. I'm using this call to determine if the block for the receipt hash is still included in the chain and older than the finalized block.

Reply:
My suspicion is that eth_getBlockByHash will return to you a block, even if it is not in the canonical chain.
Instead, if you call eth_getBlockByNumber() and then validate that its hash is same as receipt's, then it is a good check.
Because beyond the Finality Depth, the eth_getBlockByNumber() will return you the canonical block at that number.

But if you call eth_getBlockByHash(), it may return you the block, but it maynot be canonical.